### ISSUE TITLE
Add missing href="#" to links without href attributes

### ISSUE THREAD

**Issue by ndarilek** [medialize/ally.js#76]
Links without href attributes are often invisible to screen readers (or to NVDA, at any rate.) Yet many web developers misuse them. I'd like to see an API that detects missing href attributes and adds href="#" so the links are focusable and actionable. I don't know if there are any negative consequences of this but thus far I haven't found any.

I fix this where I can, but eventually other designers come along and break it again. :)

**Comment by rodneyrehm** [comment]
>  I don't know if there are any negative consequences of this but thus far I haven't found any.

`<a name="…">text</a>` is perfectly valid HTML. It provides an *anchor* the document can be navigated to in the same way `<span id="…">` does. Simply adding an `href` attribute would change semantics and behavior of that element.

If you're referring to cases such as `<a onclick="…">text</a>`, then you're right, the author has made a (grave) mistake. However, it is not possible to detect if an element has event handlers attached to it.

Your proposition operates on the assumption that `<a>` elements *require* an `href` attribute, which they do not:

> If the a element has no href attribute, then the element represents a placeholder for where a link might otherwise have been placed, if it had been relevant, consisting of just the element's contents. [HTML5.1: the a element](http://www.w3.org/TR/html51/semantics.html#the-a-element)

I understand your pain, but I don't think I have a solution for you at this time.

**Comment by ndarilek** [comment]
I definitely hear you regarding the spec, but I find that developers often do the wrong thing and leave out href attributes for links that, whether or not they *should* have href attributes (or even be links to begin with), need them so specific areas of the app can be reachable by screen readers. Conversations usually go as follows:

Me: Hey, can you please include href attributes on your links? Here's a PR that does this, and if you could do so going forward that'd be great.
Them: Will do, thanks for the PR.
Me: (a few months later, finding more links without hrefs) You've done it again. Can you please fix this and incorporate it into your CI/lint/whatever going forward?
Them: (sometimes radio silence, sometimes rinse and repeat)

If I could just point them to ally.js as a part of that first PR, it would go a long way toward me not having to submit the same PRs again and again. :)

I would be fine with it being an optional feature, but I've seen this bug enough times that, especially in an open source project where not all designers collaborate closely, I'd love to see a library handle it and maybe even throw up console logs/warnings suggesting that it be fixed.

As an example, see https://github.com/wekan/wekan Many of its links lack hrefs, and I can't even tab to them in NVDA. Leaving aside the issue that adding the hrefs still doesn't give them an aria-label, regardless of what the spec says, doing so at least lets me activate menus and interact with cards via the keyboard. I've got a PR in to fix this, but it doesn't guarantee that someone else will just introduce more of these broken links 6 months later.

Thanks.

**Comment by rodneyrehm** [comment]
> […] I'd love to see a library handle it and maybe even throw up console logs/warnings suggesting that it be fixed.

this particular request sounds like you're looking for an "accessibility auditing tool". What you're describing is *author error*. If we were to simply fix things for them, they'd not improve. What they really need is a tool (similar to jslint, jshint, eslint, …) warning them about a potential error.

ally.js is not aiming to be such a tool at this point in time. There already are a number of options available, as a quick google search reveals:

* http://accesslint.com/
* https://github.com/globant-ui/arialinter
* http://pa11y.org/
* http://www.w3.org/WAI/ER/tools/
* … among many others.

It is always hard to hear a no - and it isn't easy saying no either. ally.js is still a *very* young project and we need to carefully decide what we can spend our time on. I'd like to improve the lives of people already caring for - or ready to care for - accessibility. Once we've got *them* covered, we can start thinking about how to take on the rest of the web development world.

In other words I'd like to postpone any discussions about *automagically fixing things*. Going forward I'll tag issues like this with `wontfix` and `automagic fix` so we'll know which issues to revisit should the time come, or we change our mind about this. (@marcysutton: jump in any time if you disagree)
